src/services/FirebaseService.ts

```diff
@@ -114,16 +130,34 @@ export class FirebaseService {
    * Requests custom token for the address from Firebase function and signs in to Firebase
    */
   private async signInToFirebase(address: string): Promise<UserCredential> {
+    if (firebaseAuth.currentUser && firebaseAuth.currentUser.uid === address) {
+      return;
+    }
+
     // Signs out from Firebase in case another user was authenticated
     // (could happen when user disconnect and connect a new wallet)
     await signOut(firebaseAuth);
 
-    // Requests a custom Firebase Token used for signing in to Firebase, from our Firebase Function
-    const token = await this.createCustomToken(address);
+    const messageToSign = await this.getMessageToSign(address);
+
+    let signature: string;
+    try {
+      this.eventAggregator.publish("firebase.signature", true);
+      signature = await this.ethereumService.getDefaultSigner().signMessage(messageToSign);
+      this.eventAggregator.publish("firebase.signature", false);
+    } catch (error) {
+      this.eventAggregator.publish("firebase.signature", false);
+      this.ethereumService.disconnect({ code: 0, message: "User requested" });
+      this.eventAggregator.publish("handleFailure", "Signature was rejected. You are not authenticated");
+      return;
+    }
+
+    const token = await this.verifySignedMessageAndCreateCustomToken(address, signature);
 
-    // Firebase Authentication will be persisted in memory only
-    // that is on browser refresh the Firebase access token will be lost
-    await setPersistence(firebaseAuth, inMemoryPersistence);
+    // Firebase Authentication will be persisted in the browser storage (IndexedDB)
+    // user will be authenticated to Firebase as long as they don't clear the browser storage
+    // (or disconnect their wallet account, or switch to another account which will sign them out)
+    await setPersistence(firebaseAuth, browserLocalPersistence);
```

Comment:
Do we are about the concept of "expiration" a connected session?

Reply:
Access token expires in 60 minutes and refresh token never expires (unless revoked in certain scenarios) and it automagically refreshed the access token, if it expired or is about to expire